File: src/AzureDataFactory.TestingFramework/Models/Activities/ControlActivity/IfConditionActivity.cs
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.

using AzureDataFactory.TestingFramework.Models.Activities.Base;
using AzureDataFactory.TestingFramework.Models.Base;
using AzureDataFactory.TestingFramework.Models.Pipelines;

namespace AzureDataFactory.TestingFramework.Models;

public partial class IfConditionActivity : IIterationActivity
{
    private bool? _evaluatedExpression;
    public bool EvaluatedExpression => _evaluatedExpression ?? throw new InvalidOperationException("Expression has not been evaluated yet.");
    public override DataFactoryEntity Evaluate(PipelineRunState state)
    {
        base.Evaluate(state);

        _evaluatedExpression = Expression.Evaluate<bool>(state);

        return this;
    }

    internal override IEnumerable<PipelineActivity> EvaluateControlActivityIterations(PipelineRunState state, EvaluateActivitiesDelegate evaluateActivities)
    {
        var scopedState = state.CreateIterationScope(null);
        var activities = EvaluatedExpression ? IfTrueActivities.ToList() : IfFalseActivities.ToList();
        foreach (var activity in evaluateActivities(activities, scopedState))
            yield return activity;

        state.AddScopedActivityResultsFromScopedState(scopedState);
    }
}

File: src/python/azure_data_factory_testing_framework/models/activities/control_activities/if_condition_activity.py
from typing import Callable, Generator

from azure_data_factory_testing_framework.generated.models import (
    Activity,
    ControlActivity,
    IfConditionActivity,
)
from azure_data_factory_testing_framework.state import PipelineRunState


class IfConditionActivity:
    def evaluate(self: IfConditionActivity, state: PipelineRunState) -> IfConditionActivity:
        self.expression.evaluate(state)

        return super(ControlActivity, self).evaluate(state)

    def evaluate_control_activity_iterations(
        self: IfConditionActivity,
        state: PipelineRunState,
        evaluate_activities: Callable[[PipelineRunState], Generator[Activity, None, None]],
    ) -> Generator[Activity, None, None]:
        scoped_state = state.create_iteration_scope(None)
        activities = self.if_true_activities if self.expression.evaluated else self.if_false_activities
        for activity in evaluate_activities(activities, scoped_state):
            yield activity

        state.add_scoped_activity_results_from_scoped_state(scoped_state)
